src/common/libflux/fripp.c

```diff
@@ -0,0 +1,456 @@
+/************************************************************\
+ * Copyright 2021 Lawrence Livermore National Security, LLC
+ * (c.f. AUTHORS, NOTICE.LLNS, COPYING)
+ *
+ * This file is part of the Flux resource manager framework.
+ * For details, see https://github.com/flux-framework.
+ *
+ * SPDX-License-Identifier: LGPL-3.0
+\************************************************************/
+
+#if HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include <arpa/inet.h>
+#include <netinet/in.h>
+#include <stdbool.h>
+#include <sys/socket.h>
+#include <sys/stat.h>
+#include <syslog.h>
+#include <time.h>
+#include <unistd.h>
+
+#include "attr.h"
+#include "flog.h"
+#include "fripp.h"
+#include "reactor.h"
+
+#include "src/common/libczmqcontainers/czmq_containers.h"
+#include "src/common/libutil/iterators.h"
+
+#define FRIPP_MAX_PACKET_LEN 1440
+#define INTERNAL_BUFFSIZE 128
+#define DEFAULT_AGG_PERIOD 1.0
+
+typedef enum {
+    BRUBECK_COUNTER,
+    BRUBECK_GAUGE,
+    BRUBECK_TIMER
+} metric_type;
+
+union val {
+    ssize_t l;
+    double d;
+};
+
+struct metric {
+    union val cur;
+    union val prev;
+    bool inc;
+    metric_type type;
+};
+
+struct fripp_ctx {
+    struct sockaddr_in si_server;
+    int sock;
+    int buf_len;
+    int tail;
+    char *buf;
+    char prefix[INTERNAL_BUFFSIZE];
+
+    zhashx_t *metrics;
+    zlist_t *done;
+    flux_watcher_t *w;
+    double period;
+};
+
+bool fripp_enabled (flux_t *h, const char *metric)
+{
+    return getenv ("FLUX_FRIPP_STATSD") != NULL;
+}
+
+void fripp_set_prefix (struct fripp_ctx *ctx, const char *prefix)
+{
+    strncpy (ctx->prefix, prefix, INTERNAL_BUFFSIZE - 1);
+}
+
+static int split_packet (char *packet, int len)
+{
+    for (int i = len - 1; i >= 0; --i) {
+        if (packet[i] == '\n' || packet[i] == '\0') {
+            packet[i] = '\0';
+            return i;
+        }
+    }
+
+    return -1;
+}
+
+static int fripp_send_metrics (struct fripp_ctx *ctx)
+{
+    int len, sock_len = sizeof (ctx->si_server);
+    bool split;
+    char *packet = ctx->buf;
+
+    do {
+        split = false;
+        len = strlen (packet);
+
+        if (len > FRIPP_MAX_PACKET_LEN)
+            split = true;
+        if ((len = split_packet (packet, len > FRIPP_MAX_PACKET_LEN ?
+                FRIPP_MAX_PACKET_LEN : len)) == -1)
+            return -1;
+        if (sendto (ctx->sock, packet, len, 0,
+                    (void *) &ctx->si_server, sock_len) < 0)
+            fprintf (stderr, "packet %s dropped: %s\n", packet,
+                     strerror (errno));
```

Comment:
the fprintf to stderr should be dropped (leftover debugging?)

It's not great practice (though sometimes the only way) to print stuff on stdio from a library.  Here the entire stats collection mechanism ought to just be "best effort" IMHO.

Reply:
Should the `fprintf` on line 314 (in the `timer_cb`) also be dropped?

Reply:
Missed that one.  Yep, IMHO.